src/components/combo/combo.ts

```diff
@@ -0,0 +1,654 @@
+    this.requestUpdate();
+  }
+
+  protected override async getUpdateComplete() {
+    const result = await super.getUpdateComplete();
+    await this.toggleController.rendered;
+    return result;
+  }
+
+  /**
+   * Returns the current selection as a list of commma separated values,
+   * represented by the display key, when provided.
+   */
+  public get value() {
+    return this.selectionController.getValue(
+      Array.from(this.selectionController.selected)
+    );
+  }
+
+  @watch('value')
+  protected validate() {
+    this.updateComplete.then(() => this.reportValidity());
+  }
+
+  /** Checks the validity of the control. */
+  public reportValidity() {
+    this.invalid = this.required && !this.value;
+    return !this.invalid;
+  }
+
+  /** A wrapper around the reportValidity method to comply with the native input API. */
+  public checkValidity() {
+    return this.reportValidity();
+  }
+
+  /** Sets focus on the component. */
+  public override focus(options?: FocusOptions) {
+    this.target.focus(options);
+  }
+
+  /** Removes focus from the component. */
+  public override blur() {
+    this.target.blur();
+  }
+
+  /**
+   * Selects the options in the list by either value or key value.
+   * If not argument is provided all items will be selected.
+   * @param { T[] | Values<T>[] } items - A list of values or values as set by the valueKey.
+   */
+  public select(items?: T[] | Values<T>[], emit = false) {
+    this.selectionController.select(items, emit);
+  }
+
+  /**
+   * Deselects the options in the list by either value or key value.
+   * If not argument is provided all items will be deselected.
+   * @param { T[] | Values<T>[] } items - A list of values or values as set by the valueKey.
+   */
+  public deselect(items?: T[] | Values<T>[], emit = false) {
+    this.selectionController.deselect(items, emit);
+  }
+
+  protected handleSearchInput(e: CustomEvent) {
+    this.dataController.searchTerm = e.detail;
+  }
+
+  protected handleOpening() {
+    const args = { cancelable: true };
+    return this.emitEvent('igcOpening', args);
+  }
+
+  protected handleClosing(): boolean {
+    const args = { cancelable: true };
+    return this.emitEvent('igcClosing', args);
+  }
+
+  /** Shows the list of options. */
+  public async show(emit = false) {
+    if (this.open) return;
+    if (emit && !this.handleOpening()) return;
+    this.open = true;
+
+    await this.updateComplete;
+    emit && this.emitEvent('igcOpened');
+
+    this.list.focus();
+
+    if (!this.autofocusList) {
+      this.input.focus();
+    }
+  }
+
+  /** Hides the list of options. */
+  public async hide(emit = false) {
+    if (!this.open) return;
+    if (emit && !this.handleClosing()) return;
+    this.open = false;
+
+    await this.updateComplete;
+    emit && this.emitEvent('igcClosed');
+    this.target.focus();
+  }
+
+  /** Toggles the list of options. */
+  public toggle(emit = false) {
+    this.open ? this.hide(emit) : this.show(emit);
+  }
```

Comment:
Same for `emit` here.
These should basically set `open` to the respective value. And since `open` does not emit, neither should these. Dispatch the events from internal methods on user interaction.

Reply:
Done.